src/java/org/apache/cassandra/db/filter/IndexHints.java

```diff
@@ -0,0 +1,674 @@
+        Set<Index> result = new HashSet<>();
+        for (Index index : indexes)
+        {
+            if (includes(index))
+                result.add(index);
+        }
+        return result;
+    }
+
+    /**
+     * @param indexes a collection of indexes
+     * @return {@code true} if any of the indexes is included, {@code false} otherwise
+     */
+    public boolean includesAnyOf(Collection<Index> indexes)
+    {
+        for (Index index : indexes)
+        {
+            if (includes(index))
+                return true;
+        }
+        return false;
+    }
+
+    /**
+     * @param index an index
+     * @return {@code true} if the index is excluded, {@code false} otherwise
+     */
+    public boolean excludes(Index index)
+    {
+        return excludes(index.getIndexMetadata().name);
+    }
+
+    /**
+     * @param indexName the name of an index
+     * @return {@code true} if the index is excluded, {@code false} otherwise
+     */
+    public boolean excludes(String indexName)
+    {
+        for (IndexMetadata i : excluded)
+        {
+            if (i.name.equals(indexName))
+                return true;
+        }
+        return false;
+    }
+
+    /**
+     * @param indexes a set of indexes
+     * @return the indexes that are not excluded by these hints
+     */
+    public <T extends Index> Set<T> notExcluded(Iterable<T> indexes)
+    {
+        Set<T> result = new HashSet<>();
+        for (T index : indexes)
+        {
+            if (!excludes(index))
+                result.add(index);
+        }
+        return result;
+    }
+
+    /**
+     * Returns the best of the specified indexes that satisfies the specified filter, according to these hints first,
+     * and the index-provided selectivity second.
+     *
+     * @param indexes a collection of indexes
+     * @param filter a filter to apply to the indexes
+     * @param isContains whether the calling operation is a {@code [NOT] CONTAINS [KEY]} operation, in which case we
+     * prefer not-analyzed indexes (see CNDB-13925).
+     * @return the best of the specified indexes that satisfies these index hints and the specified filter
+     */
+    public <T extends Index> Optional<T> getBestIndexFor(Collection<T> indexes, Predicate<T> filter, boolean isContains)
+    {
+        // filter excluded and filtered indexes
+        Collection<T> candidates = filter(indexes, index -> !excludes(index) && filter.test(index));
+
+        // prefer included indexes
+        candidates = prefer(candidates, this::includes);
+
+        // if we are using a contains operator, we prefer indexes without an analyzer (see CNDB-13925)
+        if (isContains)
+            candidates = prefer(candidates, index -> !index.isAnalyzed());
+
+        // return the candidate with the best selectivity
+        return bestSelectivityIndex(candidates);
+    }
+
+    /**
+     * Returns the indexes in the specified collection of indexes that are applicable to the specified filter.
```

Comment:
It can use a simpler language:
```suggestion
     * Returns the indexes in the specified collection of indexes that satisfy the specified filter.
```